image/manifest.go

```diff
@@ -26,28 +26,301 @@ import (
+	// Check against any pre-defined restrictions.
+	if allowed, ok := sm.vsAllowed[vs.Unique()]; ok {
+		if allowed.Sequence != vs.Sequence {
+			return fmt.Errorf("sequence of %q (%d) does not match the allowed sequence (%d)",
+				vs.Unique(), vs.Sequence, allowed.Sequence)
+		}
+		if allowed.Pinned != pinned {
+			return fmt.Errorf("pinning of %q (%t) does not match the allowed pinning (%t)",
+				vs.Unique(), pinned, allowed.Pinned)
+		}
+	}
+
+	for _, sn := range vsa.Snaps() {
+		// Record only snaps that have a presence set, and a revision specified
+		if sn.Presence == asserts.PresenceInvalid {
+			continue
+		}
+		if sn.Revision <= 0 {
+			continue
+		}
+
+		// Update allowed snaps based on the validation-set.
+		if err := sm.SetAllowedSnapRevision(sn.SnapName(), sn.Revision); err != nil {
+			return err
+		}
+
+		// For book-keeping purposes we add the snap to the list of controlled
+		// snap revisions by the validation set.
+		vs.Snaps = append(vs.Snaps, sn.SnapName())
+	}
+
+	sm.vsSeeded[vs.Unique()] = vs
+	return nil
+}
+
+// AllowedRevision retrieves any specified revision rule for the snap
+// name.
+func (sm *SeedManifest) AllowedSnapRevision(snapName string) snap.Revision {
+	// TODO: Check seeded validation-sets as well.
+	if rev, ok := sm.revsAllowed[snapName]; ok {
+		return rev.Revision
+	}
+	return snap.Revision{}
+}
+
+// ValidationSetsAllowed returns the validation sets specified as allowed.
+func (sm *SeedManifest) AllowedValidationSets() []*SeedManifestValidationSet {
+	var vss []*SeedManifestValidationSet
+	for _, vs := range sm.vsAllowed {
+		vss = append(vss, vs)
+	}
+
+	// Sort for test consistency
+	sort.Slice(vss, func(i, j int) bool {
+		return vss[i].Unique() < vss[j].Unique()
+	})
+	return vss
+}
+
+func parsePinnedValidationSet(sm *SeedManifest, vs string) error {
+	acc, name, seq, err := snapasserts.ParseValidationSet(vs)
+	if err != nil {
+		return err
+	}
+	return sm.SetAllowedValidationSet(acc, name, seq, true)
+}
+
+func parseUnpinnedValidationSet(sm *SeedManifest, vs, seqStr string) error {
+	acc, name, _, err := snapasserts.ParseValidationSet(vs)
+	if err != nil {
+		return err
+	}
+	seq, err := strconv.Atoi(seqStr)
+	if err != nil {
+		return fmt.Errorf("invalid formatted validation-set sequence: %q", seqStr)
+	}
+	return sm.SetAllowedValidationSet(acc, name, seq, false)
+}
+
+func parseSnapRevision(sm *SeedManifest, sn, revStr string) error {
+	if err := snap.ValidateName(sn); err != nil {
+		return err
+	}
+
+	rev, err := snap.ParseRevision(revStr)
+	if err != nil {
+		return err
+	}
+
+	// Values that are higher than 0 indicate the revision comes from the store, and values
+	// lower than 0 indicate the snap was sourced locally. We allow both in the seed.manifest as
+	// long as the user can provide us with the correct snaps. The only number we won't accept is
+	// 0.
+	if rev.Unset() {
+		return fmt.Errorf("cannot use revision %d for snap %q: revision must not be 0", rev, sn)
```

Comment:
snap revision for %q in manifest cannot be 0 (unset)

or something like that, avoids repeating 0 twice

Reply:
That makes much more sense